discojs/src/dataset/dataset.ts

```diff
@@ -144,8 +144,48 @@ export class Dataset<T> implements AsyncIterable<T> {
     );
   }
 
-  /** Flatten chunks */
-  unbatch<U>(this: Dataset<Batched<U>>): Dataset<U> {
+  /**
+   * Create batches of size `size + 1` which overlap on one element: 
+   * the last element of one batch is the same as the first element of the next
+   * Notes:
+   * - The resulting dataset has a batch size `size`+ 1
+   * - The last batch is dropped as there are no next element to add.
+   *
+   * This method is tailored to create text sequences where each token's label is the following token. 
+   * In order to have a label for the last token of the input sequence, we include the first token
+   * of the next sequence.
+   * 
+   * @param size batch size excluding the overlapping element, at least 1
+   * @returns a dataset batch size `size + 1`
+   */
+  batchWithOverlap(size: number): Dataset<Batched<T>> {
+    if (size <= 0 || !Number.isInteger(size)) throw new Error("invalid size");
+    
+    return new Dataset(
+      async function* (this: Dataset<T>) {
+        const iter = this.batch(size)[Symbol.asyncIterator]();
+        // get the first batch
+        const firstRes = await iter.next()
+        if (firstRes.done) return
+        let currentBatch = firstRes.value
+        for (; ;) {
+          // get the next batch
+          const res = await iter.next()
+          if (res.done) break;
+          const nextBatch = res.value
+          // get the first element of the next batch
+          const nextFirstElement = nextBatch.first()
+          if (nextFirstElement === undefined) break
+          // yield the current batch with the first element of the next batch
+          yield currentBatch.concat(nextFirstElement);
+          currentBatch = nextBatch
+        }
+      }.bind(this),
+    );
+  }
+
+  /** Flatten batches/arrays of elements */
+  flat<U>(this: Dataset<Batched<U>>): Dataset<U> {
```

Comment:
`flatten` is a [more](https://immutable-js.com/docs/latest@main/List/#flatten()) [standard](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flatten) name. also, now that we changed the name, we can input `DatasetLike<U>` instead of `Batched<U>` to allow flattening `List`, arrays, other `Dataset`, …

Reply:
Because of code duplication with the constructor I created a common function to convert DatasetLike to AsyncGenerator, let me know if there was a simpler way to do it!

Reply:
ho, good idea, I was reusing the constructor but a standalone function is more readable :ok_hand: